tcpdump.c

```diff
@@ -0,0 +1,239 @@
+    error_status_t ret_status = STATUS_SUCCESS;
+    
+    CHECK(packet != NULL);
+
+    // Basic packet parser
+    // add more: ip level, tcp/udp, blah blah
+    struct ethhdr* eth_h = (struct ethhdr*)packet;
+	printf("Ethernet Header\n");
+	printf("\t|-Source Address : %.2X:%.2X:%.2X:%.2X:%.2X:%.2X\n", eth_h->h_source[0],eth_h->h_source[1],eth_h->h_source[2],
+   		eth_h->h_source[3],eth_h->h_source[4],eth_h->h_source[5]);
+	printf("\t|-Destination Address : %.2X:%.2X:%.2X:%.2X:%.2X:%.2X\n", eth_h->h_dest[0],eth_h->h_dest[1],eth_h->h_dest[2],
+   		eth_h->h_dest[3],eth_h->h_dest[4],eth_h->h_dest[5]);
+
+cleanup:
+    return ret_status;
+}
+
+static error_status_t sniff(int sock, int fd) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    struct sockaddr_ll sll = {0};
+    socklen_t sll_len = sizeof(sll);
+    unsigned char buffer[MAX_PKT_LEN] = {0};
+    ssize_t r_bytes = 0;
+
+    while (running) {
+        // 0 is for the flags field, no special flags are used
+        r_bytes = recvfrom(sock, buffer, sizeof(buffer), 0, (struct sockaddr*)&sll,
+                           &sll_len);
+        if (r_bytes == MAX_PKT_LEN) {
+            printf("Packet len too long\n");
+            continue;
+        }
+
+        // TODO: is this the best way to handle this?
+        // the problem here is that when we SIGINT recv terminates,
+        // we want to handle this without CHECKing as it goes into cleanup.
+        if (r_bytes == -1) {
+            if (errno == EINTR) {
+                continue;
+            }
+        }
+
+        CHECK(r_bytes < sizeof(buffer));
+
+        // TODO: print sockaddr info?
+        if (fd != -1) {
+            write_packet(fd, buffer, r_bytes);
+        } else {
+            handle_packet(buffer);
+        }
+    }
+
+cleanup:
+    return ret_status;
+}
+
+error_status_t my_tcpdump(const char* if_name, const char* output_file, const char* bpf) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    int raw_sock = -1;
+    int fd = -1;
+    int link_type = -1;
+    
+    CHECK_FUNC(create_raw_socket(&raw_sock));
+
+    if (if_name != NULL) {
+        CHECK_FUNC(bind_raw_socket(raw_sock, if_name));
+        // TODO: should i first check/save the interface's current state?
+        CHECK_FUNC(iface_set_promisc(raw_sock, if_name, true));
+        CHECK_FUNC(dev_get_iftype(if_name, &link_type));
+    } else {
+        // TODO: is this a good default? 
+        link_type = DLT_EN10MB; 
+    }
+
+    if (output_file != NULL) {
+      CHECK_FUNC(open_pcap_file(output_file, &fd, MAX_PKT_LEN - 1, link_type));
+    }
+    if (strlen(bpf) != 0) {
+        // the 1 is for optimizing the BPF
+        CHECK_FUNC(set_bpf(raw_sock, link_type, bpf, 1));
+    }
+
+    CHECK_FUNC(sniff(raw_sock, fd));
+
+cleanup:
+    if (if_name != NULL) {
+        // is is bad if it fails? is there a better place to put this?
+        iface_set_promisc(raw_sock, if_name, false); // Best effort.
```

Comment:
I would maybe print an error if you failed, but it's not like you could do anything about it...
Also, create a variable `set_promisc` and set it to `true` only after you set the interface to promisc, and check it to know if you should unset promisc.
Also, what if the interface was already in promisc before you started? (a bridge for example)

Reply:
so i thought that's what i should do but then when trying to find how to get the current promisc state i found this 
> IFF_PROMISC is not set,
It's not supposed to be set.
The correct way to put into promiscuous mode the device to which a
PF_PACKET socket is to do a SOL_PACKET/PACKET_ADD_MEMBERSHIP
"setsockopt()" call with PACKET_MR_PROMISC as the argument (see the
"packet(7)" man page), and that's what libpcap is doing.
The old way of directly setting IFF_PROMISC had problems - to quote the
comment at the front of "pcap-linux.c":
We have to set the interface's IFF_PROMISC flag ourselves, if
we're to run in promiscuous mode, which means we have to turn
it off ourselves when we're done; the kernel doesn't keep track
of how many sockets are listening promiscuously, which means
 it won't get turned off automatically when no sockets are
listening promiscuously. We catch "pcap_close()" and, for
interfaces we put into promiscuous mode, take them out of
promiscuous mode - which isn't necessarily the right thing to
do, if another socket also requested promiscuous mode between
the time when we opened the socket and the time when we close the socket.
With the new mechanism, the kernel *does* keep track of how many
requests for promiscuous mode there were, so that libpcap doesn't have
to turn promiscuous mode off by itself.
The kernel code appears to set the IFF_PROMISC flag in the "flags" field
of the "struct device" structure for the interface - see
"dev_set_promiscuity()" in "net/core/dev.c".
However, the code to handle SIOCGIFFLAGS doesn't return the IFF_PROMISC
flag from that field, ti returns the IFF_PROMISC flag from the "gflags"
field - see "dev_ifsioc()" in "net/core/dev.c".
"dev_change_flags()" (also in "net/core/dev.c") does appear to set the
IFF_PROMISC flag in "gflags" if the "flags" argument has it set; it then
calls "dev_set_promiscuity()".
However, the code to handle PACKET_MR_PROMISC directly calls
"dev_set_promiscuity()"; it doesn't set the "gflags" bit.
This means that only promiscuity requested by SIOCSIFFLAGS will show up
in SIOCGIFFLAGS, not promiscuity requested by PACKET_MR_PROMISC.
This may be intentional; if anybody doesn't like that behavior, they
should take it up with the Linux networking folk - I don't think libpcap
should use the deprecated SOCK_PACKET mechanism merely so that
"ifconfig" will report PROMISC on interfaces on which promiscuous
captures are being done, especially given that there are *other*
versions of libpcap that use PF_PACKET sockets as well (e.g., the
versions with Alexey Kuznetzov's patches - the Red Hat 6.1 and later,
and SuSE 6.whatever and later, distributions have those patches
applied).

I haven't found out how i can get the promisc state in another way. 
For example, bridges just literally check the flags, and hold there own state.
you can see this functionality here `br_port_set_promisc` in `net/bridge/br_if.c`.

This is still okay because it looks like the kernel handles a counter which keeps track of how many times promisc has been set and removes it when it reaches 0.
And pcap doesn't have a better implementation so, i feel comfortable leaving it as is.
I am going to add a function using the ioctl but im not going to use it.